api/pkg/controller/kubeletdnat/kubeletdnat_controller.go

```
@@ -241,55 +211,45 @@ func getInternalNodeAddress(node *corev1.Node) (string, error) {
 	return "", fmt.Errorf("no internal address found; known addresses: %v", node.Status.Addresses)
 }
 
-// getRuleFromNode determines the used kubelet address of a node
-// and creates a DnatRule from it.
-func (ctrl *Controller) getRuleFromNode(node *corev1.Node) (*DnatRule, error) {
-	if node == nil {
-		return nil, fmt.Errorf("invalid/nil node reference")
-	}
-	rule := &DnatRule{}
-	// Set matching part of the rule (original address).
-	host, err := getPreferredNodeAddress(node)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get preferred node address: %v", err)
-	}
+// getRulesForNode determines the used kubelet address of a node
+// and creates a dnatRule from it.
+func (ctrl *Controller) getRulesForNode(node *corev1.Node) []*dnatRule {
+	rules := []*dnatRule{}
+
 	port := int(node.Status.DaemonEndpoints.KubeletEndpoint.Port)
 	if port <= 0 {
 		port = provider.DefaultKubeletPort
 	}
-	rule.OriginalTargetAddress = host
-	rule.OriginalTargetPort = strconv.FormatInt(int64(port), 10)
 
-	// Set translation part of the rule (new destination)
-	//    This implements the current node-access-network translations by
-	//    changing the first two octets of the node-ip-address into the
-	//    respective two octets of the node-access-network.
 	internalIP, err := getInternalNodeAddress(node)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get internal node address: %v", err)
+		glog.Errorf("failed to get internal node address: %v", err)
```

Why does the log message say: `failed to ...` but the function returns successfully?
If `getRulesForNode` cannot create the rules for the given input because the given input is not sufficient, `getRulesForNode` should return an error.

The calling function `getDesiredRules` should instead decide if a single failure should make the whole ruleset creation fail

---

When rules for a single node cannot be created an error is logged and rule-creation continues for the other nodes. This is what is happening here. The error is logged and an empty list is returned.

If you say the decision "log but do not fail" should happen more upwards in the call-tree, then I can move it there.

---

Yes, i would definitely prefer doing the decision one level above.
Also we should include something like `could not generate rules for node foo: %v - skipping`.
That way a operator knows that logged error is not critical